File: 21558 - Nithin/week 5/785-is-graph-bipartite/785-is-graph-bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = graph.size();
        
        vector<int> colors(n, 0);
        
        for(int i = 0;i < n;i++){
            
            if(colors[i] == 0){
                colors[i] = -1;
                queue<int> q;
                q.push(i);
            
                while(!q.empty()){
                    int front = q.front();
                    q.pop();
                    
                    for(int j = 0;j < graph[front].size();j++){
                        if(colors[graph[front][j]] == colors[front])
                            return false;
                        else if(colors[graph[front][j]] == 0){
                            colors[graph[front][j]] = -1*colors[front];
                            q.push(graph[front][j]);
                        }
                        else
                            continue;
                        }
                
                    }
                }
            }
        return true;
    }
};

File: 21551-naveen-siva-kumar/week5/785-is-graph-bipartite/785-is-graph-bipartite.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int n = graph.length;
        int colors[] = new int[n];
        
        Queue<Integer> q = new LinkedList<>();
        int qEle;
        
        for (int i = 0; i < n; i++) {
            if(colors[i] == 0) {
                colors[i] = 1;
                q.offer(i);
            }             
            while (!q.isEmpty()) {
                qEle = q.poll();  //for each of the neighbour,traverse through and color again.
                
                for (int neighobour : graph[qEle]) {                       
                    if(colors[neighobour] == 0) {
                        colors[neighobour] = -colors[qEle];
                        q.offer(neighobour);
                    } 
                    else if (colors[qEle] == colors[neighobour]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
}
